### ISSUE TITLE
[New build - EXPERIMENTAL] 22w19a: New Terrain and Asteroids

### ISSUE THREAD

**Issue by tsunamayo** [tsunamayo/Starship-EVO#4695]
The asteroid technology has been completely overhauled thanks a brand new planetary terrain technology.
This terrain tech will be extended to Space Habitat and planets in upcoming updates.

It is a massive update as so many tech behind the scene has to be implemented or improved to deal with the amount of asteroids drawn.
Beside the new features the new terrain/asteroids use a different approach: it is way more generic and also leverage fully on other tech instead of being completely bespoke.
It was quite tricky to pull off as it stresses and push to the limit so many area of the game technology which had to be improved. 
In the end it is a great thing as this will benefit other sections of the game beside terrain!

=> Please note that the system size has been reduced, so any player entities will unfortunately have to be repositioned if necessary.

**New Terrain Technology:**
- Feature LOD: as you move closer finer detailed terrain patches will fade-in.
- Terrain deformation: You can deform the terrain with weapons (for now, a proper mining feature will be introduced later)
- Fully integrated with the existing brick entity rendering technology, which means GPU culling and minimum CPU impact. 
- Fully integrated with the existing physics technology, can be multi-thread for minimum CPU impact.
- Texture and terrain shape are very much wip and will be improved in a later update.
- Still a bit laggy if you zoom in fast onto terrain. It was already heavily optimized but still needs more work.

**Asteroid Technology:**
- Asteroids are now visible from much further away, the transition between sprite and actual geometry being much smoother and believable.
- Flying close the asteroid impact FPS much less compared to before (computations are multi-threaded)
- A few much larger asteroid are being spawned (several km wide)
- Generic asteroids (the smaller one, which are from a set of precomputed) also have a two levels LOD (on top of the builtin terrain LODs) 
- Current asteroid are gray rock but will be more varied in the future, and the icy ones will also come back.

**Other Improvements:**
- Culling data was optimized, resulting in less memory allocated on the GPU (saving 100mb ++) and less hiccup
- Culling data transfer to GPU improved, resulting in less hiccup during ship loading or warping.
- Transform data transfer to GPU was optimized, with expensive TRS matrix computations done on a Compute Shader, which reduce CPU load
- World physics query was also improved, which reduce CPU load when many entity are scattered through the world.
- Particle Data CPU cost improved.

Next up, proper mining!

**Comment by tsunamayo** [comment]
Build is up! Possible my hardest build to date... Given how deep I had to dig into various areas I expect a reasonably large number of regressions so please report anything!😭 Most likely rendering and physics related.

**Comment by SovietSponge** [comment]
Oh hell yeah terrain deformation! Can't wait to get mining tools, hope we get some interesting ones to work with.

**Comment by Kaiser-Indrasil** [comment]
Alrighty, massive update to tackle ladies and gents!

Here's a fresh bug report!
![image](https://user-images.githubusercontent.com/63756057/168018797-35adb521-dbb0-4a27-92cd-db72d6fa53f0.png)
When I damage an asteroid with lasers, they deform the meshes as they should but also poke a hole in them as well.

Here I'm at the edge of a deformed mesh fragment and the physics hold...
![image](https://user-images.githubusercontent.com/63756057/168019062-b3f30131-a620-48ac-b0d9-73dc52fd5a7a.png)

...until I fly in further at which point I can fully roam around the asteroid.
![image](https://user-images.githubusercontent.com/63756057/168019209-6511d3f0-4fab-4877-9fd0-1ed0b0c84ac5.png)

**Comment by Kaiser-Indrasil** [comment]
Another bug related to the previous one is, when I fly away a couple of hundred meters away from the deformed mesh, it patches up on the next LOD.
If I then get closer again, the hole is patched (in the collision mesh too, so I cannot physically get into the asteroid).

Initial:
![image](https://user-images.githubusercontent.com/63756057/168019604-53d61cb7-1902-4765-99a2-50f31f2e470f.png)
Farther LOD:
![image](https://user-images.githubusercontent.com/63756057/168019657-3c8b3944-fd6f-479e-aca7-a28e692d34b2.png)
Closer LOD once again:
![image](https://user-images.githubusercontent.com/63756057/168019717-d68a7b89-d5b5-46ed-a7d8-3afba4692abc.png)

**Comment by tsunamayo** [comment]
@Kaiser-Indrasil Yep I know it is normal at this stage, I havent spend that much time on the mining / deformation yet. But keep reporting any issues!

**Comment by tsunamayo** [comment]
Btw please also report your FPS and FPS drop when closing on asteroids (along with your CPU and how many cores you have). I want to check especially with CPU with few cores.

**Comment by ProPeach** [comment]
> Btw please also report your FPS and FPS drop when closing on asteroids (along with your CPU and how many cores you have). I want to check especially with CPU with few cores.

The only real fps drops I'm experiencing are when I zoom into the asteroids very quickly, but it sounds like you're already aware of that issue. It's a very brief, but very heavy stutter. I'm running an Ryzen 5 3600

https://user-images.githubusercontent.com/27857711/168044629-865687c6-ae6b-4444-a820-81dcb6d9f570.mp4


https://user-images.githubusercontent.com/27857711/168044906-7ae7f20b-5315-4ceb-b6fd-14445bd1e5f2.mp4

Fantastic implementation in general, the further render distances for the sprites is incredible!

**Comment by ProPeach** [comment]
![image](https://user-images.githubusercontent.com/27857711/168051143-0f19fe00-a6bf-418c-93f6-060c12752a00.png)
I am sometimes finding a chunk that refuses to load no matter the orientation of the camera, sadly it's not easily reproduceable

Also, are you still looking at tweaking the LOD distances for asteroids? Even the next closest asteroids seem to be on their lowest LOD
![image](https://user-images.githubusercontent.com/27857711/168051641-ace0ce85-57f5-461f-947b-c9815484b503.png)

**Comment by zorozeenee** [comment]
Crashes my game if i stay around asteroids too long, already submitted a report

**Comment by BigBadKangaroo** [comment]
All transparent blocks are completely invisible:
![20220512212445_1](https://user-images.githubusercontent.com/81495956/168153949-1b6754fd-b7a3-477f-9029-324cc09b35ad.jpg)
Vent-block misses it's grid texture:
![20220512212533_1](https://user-images.githubusercontent.com/81495956/168153964-c8ffc88e-e018-4a0b-a9ce-4a488e6f9c5e.jpg)
Blocks with sprites in them such as buttons/ sequencer gates / ship computers / gravity & shipyard & starter block etc are "empty" (no sprite in them)![20220512212826_1](https://user-images.githubusercontent.com/81495956/168153976-f1d7ad4e-e044-4468-9f18-0d4f1fe93d0e.jpg)
![20220512212842_1](https://user-images.githubusercontent.com/81495956/168153990-dd029475-656a-4e61-bd9f-80404c857bbc.jpg)

And a ton of flickering blocks / doesn't matter if wedge or regular

Near asteroids I still get around 60fps with no issue but the asteroids are visible through a planet (but not everytime)!:
![20220512214220_1](https://user-images.githubusercontent.com/81495956/168155689-a7fc180c-d58e-4929-bd1b-80d6782dc030.jpg)

**Comment by Uncle-Ulty** [comment]
Some asteroids are much bigger than others, is it supposed to be ok?

![image](https://user-images.githubusercontent.com/73253472/168169232-5398bfeb-f85b-4f77-ae79-b78702578acd.png)


serrated edge was planned, or is it some polygon-less optimization algorithm?

![image](https://user-images.githubusercontent.com/73253472/168169495-e1e4ff38-e227-41e6-9ec2-c80e7f9ad700.png)

Also, the game froze with no apparent cause (I can't reproduce the problem)
log:
[Uploading Player.log…]()
blueprint file:
[Excelência - MR Interceptor.zip](https://github.com/tsunamayo/Starship-EVO/files/8682643/Excelencia.-.MR.Interceptor.zip)

**Comment by tsunamayo** [comment]
> Crashes my game if i stay around asteroids too long, already submitted a report

@zorozeenee Hi, how long was it? Also what is your CPU? Thanks

**Comment by tsunamayo** [comment]
@Uncle-Ulty Yes the big bois are a feature. For the serrated edge unfortunately it is an artifact of the meshisation algo (marching cube). 
Also could you join the player.log, it seems it had some issue during uploading (if you still have it). What is your CPU btw?

**Comment by tsunamayo** [comment]
@BigBadKangaroo Hi, do you reproduce this right away or after spawning/unspawning ship (the brick issue). Thanks

**Comment by tsunamayo** [comment]
@BigBadKangaroo So for the planet bug, could you send me your save (you seems to be in another system). It might be because this planet is smaller.

**Comment by Uncle-Ulty** [comment]
reupload:

[Player.log](https://github.com/tsunamayo/Starship-EVO/files/8683911/Player.log)
[HIGHWIND - Final Fantasy VII Airship.zip](https://github.com/tsunamayo/Starship-EVO/files/8683913/HIGHWIND.-.Final.Fantasy.VII.Airship.zip)
[Excelência - MR Interceptor.zip](https://github.com/tsunamayo/Starship-EVO/files/8683914/Excelencia.-.MR.Interceptor.zip)

**Comment by BigBadKangaroo** [comment]
@tsunamayo the world is called Thumbnail ... a strange name I know, I created it a week ago.
[Thumbnail.zip](https://github.com/tsunamayo/Starship-EVO/files/8691294/Thumbnail.zip)

small update: I tested the same thing in a fresh new world. The bug I reported doesn't happen anymore, but in a new world you can still see the large asteroids. You can only see the small once if you get really close to the planet:
![20220513230512_1](https://user-images.githubusercontent.com/81495956/168390497-0434cd3c-ac0c-4add-86e6-054abf3befef.jpg)
![20220513231100_1](https://user-images.githubusercontent.com/81495956/168390511-f4ecbf49-5e8d-474f-bd86-1e5d52f6bf3e.jpg)

**Comment by tsunamayo** [comment]
@BigBadKangaroo all the flickering issues have been fixed in the latest 22w21a. Thanks!

**Comment by tsunamayo** [comment]
@Kaiser-Indrasil Btw the mining issues you reported are finally dealt with in the latest update! With some more improvements for the upcoming one. Cheers

**Comment by Kaiser-Indrasil** [comment]
> @Kaiser-Indrasil Btw the mining issues you reported are finally dealt with in the latest update! With some more improvements for the upcoming one. Cheers

Thanks, I've just noticed! I just hadn't had time to play in the last few weeks. Great work!
